aggregator_core/src/datastore.rs

```diff
@@ -3758,20 +3799,27 @@ impl<C: Clock> Transaction<'_, C> {
                     batch_identifier, aggregation_param, state, outstanding_aggregation_jobs,
                     client_timestamp_interval
                 FROM batches
-                WHERE task_id = (SELECT id FROM tasks WHERE task_id = $1)",
+                JOIN tasks ON tasks.id = batches.task_id
+                WHERE tasks.task_id = $1
+                  AND UPPER(client_timestamp_interval) >= COALESCE($2::TIMESTAMP - tasks.report_expiry_age * '1 second'::INTERVAL, '-infinity'::TIMESTAMP)",
             )
             .await?;
-        self.query(&stmt, &[/* task_id */ task_id.as_ref()])
-            .await?
-            .into_iter()
-            .map(|row| {
-                let batch_identifier =
-                    Q::BatchIdentifier::get_decoded(row.get("batch_identifier"))?;
-                let aggregation_parameter =
-                    A::AggregationParam::get_decoded(row.get("aggregation_param"))?;
-                Self::batch_from_row(*task_id, batch_identifier, aggregation_parameter, row)
-            })
-            .collect()
+        self.query(
+            &stmt,
+            &[
+                /* task_id */ task_id.as_ref(),
+                /* now */ &self.clock.now().as_naive_date_time()?,
```

Comment:
Remark: I do wish there was some way to dedupe this pattern, but I haven't thought of anything that would be more complexity than its worth.

Reply:
Agreed. 👍🏻